beacon-chain/p2p/crawler.go

```diff
@@ -0,0 +1,387 @@
+// crawler.go implements a discovery-based peer crawler to maintain a cache
+// of peers that support specific gossipsub topics.
+package p2p
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var crawlTimeout = 20 * time.Second
+var cleanupInterval = 5 * time.Minute
+
+type peerNode struct {
+	id     enode.ID
+	node   *enode.Node
+	peerID peer.ID
+	topics []string
+}
+
+// Crawler periodically crawls the discovery network for peers and maintains a
+// mapping of topics to the peers that support them and their ENRs.
+// TODO: Prevent unbounded growth of the topicToPeers map.
+type Crawler struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// Discovery interface for finding peers
+	dv5 Listener
+
+	topicExtractor func(ctx context.Context, node *enode.Node) ([]string, error)
+
+	// Reference to the p2p service for peer filtering and scoring
+	service *Service
+
+	crawlInterval time.Duration
+	crawlTimeout  time.Duration
+
+	wg sync.WaitGroup
+
+	mu              sync.RWMutex
+	started         bool                             // track if crawler has been started
+	peerNodes       map[enode.ID]*peerNode           // enode.ID -> peerNode record
+	topicToPeers    map[string]map[enode.ID]struct{} // topic -> set of enode.IDs
+	peerIDToEnodeID map[peer.ID]enode.ID             // libp2p peer.ID -> discv5 enode.ID
+}
+
+// newCrawler creates a new peer crawler.
+func newCrawler(ctx context.Context, dv5 Listener, service *Service, interval time.Duration) (*Crawler, error) {
+	if dv5 == nil {
+		return nil, errors.New("discovery client is required")
+	}
+	if service == nil {
+		return nil, errors.New("p2p service reference is required")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	return &Crawler{
+		ctx:             ctx,
+		cancel:          cancel,
+		dv5:             dv5,
+		service:         service,
+		crawlInterval:   interval,
+		crawlTimeout:    crawlTimeout,
+		peerNodes:       make(map[enode.ID]*peerNode),
+		topicToPeers:    make(map[string]map[enode.ID]struct{}),
+		peerIDToEnodeID: make(map[peer.ID]enode.ID),
+	}, nil
+}
+
+// Start begins the periodic crawling of the network.
+func (c *Crawler) Start(topicExtractor func(ctx context.Context, node *enode.Node) ([]string, error)) error {
+	if topicExtractor == nil {
+		return errors.New("topic extractor cannot be nil")
+	}
+	c.topicExtractor = topicExtractor
+
+	// Mark crawler as started
+	c.mu.Lock()
+	if c.started {
+		c.mu.Unlock()
+		return errors.New("crawler already started")
+	}
+	c.started = true
+	c.mu.Unlock()
+
+	log.Info("Starting peer crawler")
+	c.wg.Add(3)
+	go c.run()
+	go c.logStats()
+	go c.cleanupPeers()
+	return nil
+}
```

Comment:
2 things here:
- [sync.OnceValue](https://pkg.go.dev/sync#OnceValue) could reduce the complexity of creating this singleton behavior. (note that you currently have a race around setting topicExtractor outside the mutex)
- a type for `topicExtractor` would make the func sig easier on the eyes
```suggestion
// topicExtractor is a callback function that returns the set of all topics that it can determine
// a given peer is subscribed to based on information in the ENR record
type topicExtractor func(ctx context.Context, node *enode.Node) ([]string, error)

// Start spawns the Crawler runloops, which will continue in a set of goroutines
// until the parent context set in `newCrawler` is canceled. It uses sync.Once to
// guarantee that it only spawns one set of goroutines.
func (c *Crawler) Start(tex topicExtractor) error {
	return sync.OnceValue(func() error {
		return c.spawn(tex)
	})()
}

func (c *Crawler) spawn(tex topicExtractor) error {
	if tex == nil {
		return errors.New("topic extractor cannot be nil")
	}
	c.topicExtractor = tex
	log.Info("Starting peer crawler")
	c.wg.Go(c.run)
	c.wg.Go(c.logStats)
	c.wg.Go(c.cleanupPeers)
	return nil
}
```